File: Evaluate Division/evaluate-division.py
class Solution(object):
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """
        def build_graph(equations, values):
            graph = {}
            for i, (dividend, divisor) in enumerate(equations):
                if dividend not in graph:
                    graph[dividend] = {}
                if divisor not in graph:
                    graph[divisor] = {}
                graph[dividend][divisor] = values[i]
                graph[divisor][dividend] = 1.0 / values[i]
            return graph
        
        def dfs(dividend, divisor, visited, graph, temp):
            if dividend not in graph or divisor not in graph:
                return -1.0
            if dividend == divisor:
                return temp
            visited.add(dividend)
            for neighbor, value in graph[dividend].items():
                if neighbor not in visited:
                    result = dfs(neighbor, divisor, visited, graph, temp * value)
                    if result != -1.0:
                        return result
            return -1.0
        
        graph = build_graph(equations, values)
        
        results = []
        for dividend, divisor in queries:
            results.append(dfs(dividend, divisor, set(), graph, 1.0))
        
        return results

File: Evaluate Division/evaluate-division.java
class Solution {
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        Map<String, Map<String, Double>> graph = formGraph(equations, values);
        double[] results = new double[queries.size()];

        for (int i = 0; i < queries.size(); i++) {
            String dividend = queries.get(i).get(0);
            String divisor = queries.get(i).get(1);

            if (!graph.containsKey(dividend) || !graph.containsKey(divisor)) {
                results[i] = -1.0;
            } else {
                Set<String> visited = new HashSet<>();
                double[] ans = new double[]{-1.0};
                dfs(dividend, divisor, graph, visited, ans, 1.0);
                results[i] = ans[0];
            }
        }
        return results;
    }

    private Map<String, Map<String, Double>> formGraph(List<List<String>> equations, double[] values) {
        Map<String, Map<String, Double>> graph = new HashMap<>();
        for (int i = 0; i < equations.size(); i++) {
            String dividend = equations.get(i).get(0);
            String divisor = equations.get(i).get(1);

            graph.putIfAbsent(dividend, new HashMap<>());
            graph.putIfAbsent(divisor, new HashMap<>());

            graph.get(dividend).put(divisor, values[i]);
            graph.get(divisor).put(dividend, 1.0 / values[i]);
        }
        return graph;
    }

    private void dfs(String dividend, String divisor, Map<String, Map<String, Double>> graph, 
                     Set<String> visited, double[] ans, double temp) {
        if (visited.contains(dividend)) return;
        visited.add(dividend);

        if (dividend.equals(divisor)) {
            ans[0] = temp;
            return;
        }

        for (Map.Entry<String, Double> neighbor : graph.get(dividend).entrySet()) {
            dfs(neighbor.getKey(), divisor, graph, visited, ans, temp * neighbor.getValue());
        }
    }
}
